crates/node-maintainer/src/maintainer.rs

```diff
@@ -120,113 +131,177 @@ impl NodeMaintainer {
+                        if satisfies {
+                            names.insert(name);
+                            continue;
+                        }
+
+                        // Otherwise, we have to fetch package metadata to
+                        // create a new node (which we'll place later).
+                        in_flight_count.fetch_add(1, atomic::Ordering::SeqCst);
+                        deps.push(node_dep);
                         names.insert(name);
                     }
-                }
 
-                package_groups.push(futures::stream::iter(packages));
-            }
+                    // Release mutex so that code below doesn't dead lock
+                    drop(graph);
 
-            // Order doesn't matter here: each node name is unique, so we
-            // don't have to worry about races messing with placement.
-            let mut stream = futures::stream::iter(package_groups.drain(..))
-                .flatten()
-                .buffer_unordered(30);
-            while let Some((package, requested, dep_type, dependent_idx)) = stream.next().await {
-                let package = package?;
-                let name = UniCase::new(package.name().to_string());
-                if Self::satisfy_dependency(
-                    &mut self.graph,
-                    dependent_idx,
-                    &dep_type,
-                    &name,
-                    &requested,
-                )? {
-                    continue;
-                }
-                q.push_back(Self::place_child(
-                    &mut self.graph,
-                    dependent_idx,
-                    package,
-                    dep_type,
-                )?);
-            }
+                    for node_dep in deps {
+                        dep_sender.unbounded_send(node_dep)?;
+                    }
+
+                    if in_flight_count.fetch_sub(1, atomic::Ordering::SeqCst) == 1 {
+                        idx_sender.close_channel();
+                        dep_sender.close_channel();
+                    }
 
-            // We sort the current queue so we consider more shallow
-            // dependencies first, and we also sort alphabetically.
-            q.make_contiguous().sort_by(|a_idx, b_idx| {
-                let a = &self.graph[*a_idx];
-                let b = &self.graph[*b_idx];
-                match a.depth(&self.graph).cmp(&b.depth(&self.graph)) {
-                    Ordering::Equal => a.package.name().cmp(b.package.name()),
-                    other => other,
+                    Ok::<(), NodeMaintainerError>(())
                 }
-            })
-        }
+            });
+
+        futures::stream::select(
+            // Fetch dependencies at high concurrency and unordered because
+            // we sort the `idx_queue` by depth/name on every push anyway.
+            dep_receiver.buffer_unordered(100),
```

Comment:
btw, funny side-effect of this is that no more than 100 concurrent requests to the registry are in-flight at the moment!